### ISSUE TITLE
Undefined method error w/ user_ulimit on Chef 12

### ISSUE THREAD

**Issue by jeffbyrnes** [bmhatfield/chef-ulimit#57]
### Cookbook version
bmhatfield/chef-ulimit@370b4b0

### Chef-client version
12.21.4

### Platform Details
Ubuntu 16.04

### Scenario:
Use the `user_ulimit` resource on Chef 12

### Steps to Reproduce:
1. Add a `user_ulimit` resource to a cookbook
2. Attempt to converge using Chef 12

### Expected Result:
Successful convergence

### Actual Result:
Failure:

```
================================================================================
Recipe Compile Error in /tmp/kitchen/cache/cookbooks/ulimit/providers/domain.rb
================================================================================

NoMethodError
-------------
undefined method `action' for Chef::Provider::UlimitRule:Class

Cookbook Trace:
---------------
  /tmp/kitchen/cache/cookbooks/ulimit/providers/rule.rb:8:in `<class:UlimitRule>'
  /tmp/kitchen/cache/cookbooks/ulimit/providers/rule.rb:1:in `<top (required)>'
  /tmp/kitchen/cache/cookbooks/ulimit/providers/domain.rb:1:in `require_relative'
  /tmp/kitchen/cache/cookbooks/ulimit/providers/domain.rb:1:in `class_from_file'

Relevant File Content:
----------------------
/tmp/kitchen/cache/cookbooks/ulimit/providers/rule.rb:

  1:  class Chef::Provider::UlimitRule < Chef::Provider
  2:    def load_current_resource
  3:      new_resource.domain new_resource.domain.domain_name if new_resource.domain.is_a?(Chef::Resource)
  4:      node.run_state[:ulimit] ||= Mash.new
  5:      node.run_state[:ulimit][new_resource.domain] ||= Mash.new
  6:    end
  7:
  8>>   action :create do # ~FC017
  9:      node.run_state[:ulimit][new_resource.domain][new_resource.item] ||= Mash.new
 10:      node.run_state[:ulimit][new_resource.domain][new_resource.item][new_resource.type] = new_resource.value
 11:    end
 12:
 13:    action :delete do
 14:      # NOOP
 15:    end
 16:  end
 17:
```

This works well under Chef 13.

**Comment by ziggythehamster** [comment]
I think `Chef::Provider` doesn't have an `action` method in Chef 12.

**Comment by ziggythehamster** [comment]
c83088e46876479ce4f919aeacfca708f3322e93 works for me if someone wants to bisect

**Comment by jeffbyrnes** [comment]
I’ll give it a shot.

**Comment by jeffbyrnes** [comment]
Alrighty, got some results:

```
35e93a08774d1f187a87b1b0c5f2e0ecfd3081fe is the first bad commit
commit 35e93a08774d1f187a87b1b0c5f2e0ecfd3081fe
Author: Tim Smith <tsmith@chef.io>
Date:   Wed Jul 26 19:43:49 2017 -0700

    Properly extend Chef::Provider not Chef::Resource

    This is how we get use_inline_resources to work

    Signed-off-by: Tim Smith <tsmith@chef.io>

:040000 040000 3364387a691718bf7de11f7a9a6abfc3d699b876 d0644b5cb86fde7d50a53719001cce2ee2d3cd9f M	providers
:040000 040000 15171ca883507395046b5049a63c6b967f87a184 8b0c6cf026e549ca544fd42502106184dd2fbdcb M	recipes
```

**Comment by jeffbyrnes** [comment]
I did also see an issue w/ `use_inline_resources`, but ignored that to focus on this issue. That turns out to be good, considering what this commit’s message says 😄

**Comment by jeffbyrnes** [comment]
Well, I dug into this a bit, and… well, things are rather confusing to me. Some questions:

* If these are Chef 12.5-style Custom Resources, why are there still separate provider & resource files?
* Why are these written library-style, instead of using the DSL?
* Why are there actions on the resources _and_ the providers of the `ulimit_domain` and `ulimit_rule` resources?
* Would it make sense to, instead of having arbitrary `rule` blocks on a `ulimit_domain` resource, to have an array of hashes? E.g., instead of this:

```ruby
ulimit_domain 'my_user' do
  rule do
    item :nofile
    type :hard
    value 10000
  end
  rule do
    item :nofile
    type :soft
    value 5000
  end
end
```

we had this:

```ruby
ulimit_domain 'my_user' do
  rules([{
    item: :nofile
    type: :hard
    value: 10000
  ],
  [
    item: :nofile
    type: :soft
    value: 5000
  ]
end
```

We could simplify a ton of the logic inside `ulimit_domain`, and perhaps do away completely with the `load_current_resource` stuff (which should probably be refactored to be `load_current_value`) on both resources, since, in the end, a `template` is what we do, and that resource has its own idempotency we can depend upon.

**Comment by scalp42** [comment]
@tas50 any chance you could weigh in? @jeffbyrnes making some good points :)

**Comment by tas50** [comment]
This should be fixed in the 1.0 release that's up on the Supermarket now. We have kitchen tests now and everything is green.
